feat: support gatewayAPI

hkmdxlftjf opened the pull request:
feat: https://github.com/grafana/grafana-operator/issues/2026

Hello, 

I’ve extended the `grafana.spec` to include an `httpRoute` definition, for example:

```yaml
spec:
  security:
    admin_password: secret
    admin_user: root
  httpRoute:
    spec:
      parentRefs:
      - group: gateway.networking.k8s.io
        kind: Gateway
        name: eg
      rules:
      - matches:
        - path:
            type: PathPrefix
            value: /
```

This configuration creates an HTTPRoute resource that binds the Grafana instance to the specified Gateway.

Additionally, the httpRoute spec looks like this:

```yaml
spec:
  parentRefs:
  - group: gateway.networking.k8s.io
    kind: Gateway
    name: eg
  rules:
  - backendRefs:
    - group: ""
      kind: Service
      name: grafana-service
      namespace: default
      port: 3000
      weight: 1
    matches:
    - path:
        type: PathPrefix
        value: /
```

With this in place, the HTTPRoute will route traffic (via the Gateway) to the Grafana service on port 3000.

If there’s anything in the implementation that seems unreasonable or incorrect, please let me know.

Thank you!
Best regards,
Sure

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/grafana/grafana-operator?pullRequest=2273) <br/>All committers have signed the CLA.

hkmdxlftjf commented:
Hi @theSuess, could you please take a look when you have time? 🙏

weisdd commented on controllers/reconcilers/grafana/http_route_reconciler.go:
```diff
@@ -0,0 +1,300 @@
+
+// labelsSatisfyMatchExpressions checks if a given label set satisfies
+// a list of Kubernetes label selector requirements.
+func labelsSatisfyMatchExpressions(labels map[string]string, matchExpressions []metav1.LabelSelectorRequirement) bool {
```
Could you explain this change? Specifically, why it was added here?

hkmdxlftjf replied:
Because there can be multiple listeners in the `gateway` , and each listener may ignore certain route rules through [allowedRoutes](https://gateway-api.sigs.k8s.io/reference/spec/#allowedroutes) — such as by [namespace](https://gateway-api.sigs.k8s.io/reference/spec/#fromnamespaces) or [labelSelector](https://gateway-api.sigs.k8s.io/reference/spec/#fromnamespaces). When preferIngress is set to true, we need to return an available URL. The `labelsSatisfyMatchExpressions` check ensures that the current listener accepts this `HTTPRoute`, guaranteeing its availability. Additionally, this piece of code was inspired by the implementation here: https://github.com/grafana/grafana-operator/blob/master/controllers/controller_shared.go#L189

weisdd replied:
Could you, please, still answer my question on why it was added in the first place? (and why only one comment was different) - It'll help me to understand the reasoning behind your PR better.

weisdd replied:
Essentially, I'm trying to make sure that we do not accept any unreasonable changes introduced by an LLM that you might be using.

As for moving the function, I'll comment on that separately.

hkmdxlftjf commented on controllers/reconcilers/grafana/http_route_reconciler.go:
```diff
@@ -0,0 +1,300 @@
+
+// getMatchListener tries to find a Gateway listener that matches the HTTPRoute’s
+// namespace and hostname constraints, according to the Gateway API rules.
+func (r *HTTPRouteReconciler) getMatchListener(ctx context.Context, httpRoute *v2.HTTPRoute, gw *v2.Gateway) *v2.Listener {
```
Sorry, I didn’t quite understand your question.

If you mean why the parameter uses gateway instead of gatewaySpec, it’s because in the [Gateway API From field](https://gateway-api.sigs.k8s.io/reference/spec/#fromnamespaces)
, when From is set to Same, we need to check whether the HTTPRoute and the Gateway are in the same namespace.

If you mean why we need to query the Gateway, it’s because the HTTPRoute alone doesn’t provide enough information to construct an accessible URL.

weisdd replied:
That's a valid scenario, but I was rather thinking that the whole support for operator <-> grafana interaction through an `HTTPRoute` should not be implemented as it overcomplicates code for little gains. - If a Grafana instance is in the same cluster, why wouldn't someone allow the operator to communicate with Grafana directly?

weisdd replied:
If an instance is in another cluster, it's a so-called [external Grafana](https://grafana.github.io/grafana-operator/docs/examples/grafana/external_grafana/readme/). In this case, services, configmaps, ingresses, and many other objects are not reconciled.

Ingresses are special in a sense that they are easy to implement (all information is contained within a single object). And one of the key reasons to support them was to be able to reach an "internal" grafana instance (external instances weren't supported till v5) from the operator running locally (outside of cluster, just in local dev environment - no need to build an image).

weisdd commented on go.mod:
```diff
@@ -102,7 +104,7 @@ require (
 	golang.org/x/sync v0.17.0 // indirect
 	golang.org/x/tools v0.36.0 // indirect
-	gopkg.in/evanphx/json-patch.v4 v4.12.0 // indirect
+	gopkg.in/evanphx/json-patch.v4 v4.13.0 // indirect
```
Any changes in the dependencies not directly related to the PR should be reverted.

weisdd commented on controllers/reconcilers/grafana/http_route_reconciler.go:
```diff
@@ -0,0 +1,300 @@
+	}
+
+	// Assign AdminURL if ingress is preferred
+	if cr.PreferIngress() {
```
`Ingress` and `HTTPRoute` are separate resources, so I think there's a conflicting logic here.

weisdd replied:
But if both `Ingress` and `HTTPRoute` are defined, there would be a conflicting back-and-forth change between `Ingress` and `HTTPRoute` reconcilers, right?

weisdd commented on controllers/model/grafana_resources.go:
```diff
@@ -11,6 +11,7 @@ import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/runtime"
 	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	v2 "sigs.k8s.io/gateway-api/apis/v1"
```
Please, use consistent naming for imports (e.g. `gwapiv1` like in another file).

hkmdxlftjf commented:
I have removed the OperatorStageHTTPRoute and added unit tests for getMatchListener.  #If decide not to keep getMatchListener, I will remove it.

hkmdxlftjf commented:
> Things that also need to be covered in the PR:
> 
> * the operator must be able to function when the Gateway API CRD(s) is not present;
> * e2e-tests need to be extended to cover cases where the CRD IS NOT present and when it IS present;
> * controller-runtime needs to be configured to cache the HTTPRoute objects (check [this](https://github.com/grafana/grafana-operator/blob/4f7a236b245474235c4f93471bb04158e2039381/main.go#L226) for reference).

Hi, I’ve implemented these changes based on your suggestions.

In the e2e tests, I added case 17, which covers deploying Grafana when the `CRD` is not present — this works correctly as long as `grafana.spec.httpRoute == nil`.
I also added a test for deploying Grafana with the `CRD` present (but without any actual Gateway implementation), which also works as expected.
Additionally, caching has been configured for the `HTTPRoute` objects.

weisdd commented on main.go:
```diff
@@ -231,6 +233,7 @@ func main() { // nolint:gocyclo
 			&corev1.PersistentVolumeClaim{}: cacheLabelConfig,
 			&corev1.ConfigMap{}:             cacheLabelConfig, // Matching just labeled ConfigMaps and Secrets greatly reduces cache size
 			&corev1.Secret{}:                cacheLabelConfig, // Omitting labels or supporting custom labels would require changes in Grafana Reconciler
+			&gwapiv1.HTTPRoute{}:            cacheLabelConfig,
```
As it relies on a CRD that might not be present, I think the correct implementation would be to have a discovery check like for OpenShift Routes (check how `isOpenShift` is calculated).
With Routes, we've seen the operator panicking in non-OpenShift environments, that's why Route caching is hidden behind the `isOpenShift` check.

weisdd commented on go.mod:
```diff
@@ -188,6 +190,6 @@ require (
 	k8s.io/kube-openapi v0.0.0-20250910181357-589584f1c912 // indirect
 	k8s.io/utils v0.0.0-20251002143259-bc988d571ff4
-	sigs.k8s.io/json v0.0.0-20241014173422-cfa47c3a1cc8 // indirect
+	sigs.k8s.io/json v0.0.0-20250730193827-2d320260d730 // indirect
```
Any reason to make changes in deps here?

theSuess commented:
Thank you for all the work you put into this! We've decided to combine both this PR and #2283  into #2293 to simplify reviewing and avoiding duplicate work.

I've added both you and @lexfrei as co-authors on this PR, so your contribution will be properly credited.

We'll still need to decide on how to handle `PreferIngress` in combination with the GatewayAPI resources so your input in #2284 is appreciated!
